in_toto/keylib.go

```diff
@@ -2,220 +2,330 @@ package in_toto
-signable buffer provided. It returns an error if the underlying signing library
-fails.
-*/
-func GenerateEd25519Signature(signable []byte, key Key) (Signature, error) {
+GenerateSignature will automatically detect the key type and sign the signable data
+with the provided key. If everything goes right GenerateSignature will return
+a for the key valid signature and err=nil. If something goes wrong it will
+return an not initialized signature and an error. Possible errors are:
 
-	var signature Signature
+	* ErrNoPEMBlock
+	* ErrUnsupportedKeyType
 
-	seed, err := hex.DecodeString(key.KeyVal.Private)
-	if err != nil {
-		return signature, err
+*/
+func GenerateSignature(signable []byte, key Key) (Signature, error) {
+	var signature Signature
+	var signatureBuffer []byte
+	// The following switch block is needed for keeping interoperability
+	// with the securesystemslib and the python implementation
+	// in which we are storing RSA keys in PEM format, but ed25519 keys hex encoded.
+	switch key.KeyType {
+	case "rsa":
+		keyReader := strings.NewReader(key.KeyVal.Private)
+		pemBytes, err := ioutil.ReadAll(keyReader)
+		if err != nil {
+			return signature, err
+		}
+		// TODO: There could be more key data in _, which we silently ignore here.
+		// Should we handle it / fail / say something about it?
+		data, _ := pem.Decode(pemBytes)
+		if data == nil {
+			return signature, ErrNoPEMBLock
+		}
+		parsedKey, err := ParseKey(data.Bytes)
+		if err != nil {
+			return signature, err
+		}
+		hashed := sha256.Sum256(signable)
+		// We use rand.Reader as secure random source for rsa.SignPSS()
+		signatureBuffer, err = rsa.SignPSS(rand.Reader, parsedKey.(*rsa.PrivateKey), crypto.SHA256, hashed[:],
+			&rsa.PSSOptions{SaltLength: sha256.Size, Hash: crypto.SHA256})
+		if err != nil {
+			return signature, err
+		}
+	case "ed25519":
+		seed, err := hex.DecodeString(key.KeyVal.Private)
+		if err != nil {
+			return signature, err
+		}
+		// Note: We can directly use the key for signing and do not
+		// need to use ed25519.NewKeyFromSeed().
+		signatureBuffer = ed25519.Sign(seed, signable)
+	default:
+		return signature, fmt.Errorf("%w: %s", ErrUnsupportedKeyType, key.KeyType)
 	}
-	privkey := ed25519.NewKeyFromSeed(seed)
-	signatureBuffer := ed25519.Sign(privkey, signable)
-
 	signature.Sig = hex.EncodeToString(signatureBuffer)
 	signature.KeyId = key.KeyId
-
 	return signature, nil
 }
+
+func VerifySignature(key Key, sig Signature, unverified []byte) error {
+	switch key.KeyType {
+	case "rsa":
+		// Create rsa.PublicKey object from DER encoded public key string as
+		// found in the public part of the keyval part of a securesystemslib key dict
+		keyReader := strings.NewReader(key.KeyVal.Public)
+		pemBytes, err := ioutil.ReadAll(keyReader)
+		if err != nil {
+			return err
+		}
+		// TODO: There could be more key data in _, which we silently ignore here.
+		// Should we handle it / fail / say something about it?
+		data, _ := pem.Decode(pemBytes)
```

Comment:
I see this code is repeated. I'm also curious to know if ed25519 would work with pem.Decode?

Reply:
I'm starting to think this module could be split in keytype modules + an overarching module to plug the right submodule after key detection....

Reply:
Yep, this code is repeated, but still slightly different. It's again because of our RSA key handling.
We need to decode the pem block for RSA, because we store RSA keys as PEM blocks in our key objects.
For ed25519 we can't use the pem.Decode() method because we are not dealing with pem blocks here, we are dealing with hex encoded key byte strings. Not sure what you mean with modules, but I could put the part in functions, but i have no idea if this would make sense if we call this function only at one place.

Reply:
This can just be 

``` go
data, _ := pem.Decode([]byte(key.KeyVal.Public))
```

Reply:
Thanks a lot! @icholy I have applied your patch :100: